R/differential_abundance_analysis.R

```diff
@@ -0,0 +1,183 @@
+#' \code{FindMarkers} for more information.
+#'
+#' @rdname RunDAA
+#' @method RunDAA Seurat
+#'
+#' @examples
+#' library(pixelatorR)
+#' library(dplyr)
+#' library(SeuratObject)
+#'
+#' pxl_file <- system.file("extdata/five_cells",
+#'   "five_cells.pxl",
+#'   package = "pixelatorR"
+#' )
+#' # Seurat objects
+#' se <- ReadMPX_Seurat(pxl_file)
+#' se <- merge(se, rep(list(se), 9), add.cell.ids = LETTERS[1:10])
+#' se$sample <- c("T", "C", "C", "C", "C") %>% rep(times = 10)
+#' se <- Seurat::NormalizeData(se, normalization.method = "CLR", margin = 2)
+#'
+#' # Run DPA
+#' daa_markers <- RunDAA(se,
+#'   contrast_column = "sample",
+#'   targets = "T", reference = "C"
+#' )
+#' daa_markers
+#'
+#' @export
+#'
+RunDAA.Seurat <- function(
+  object,
+  contrast_column,
+  reference,
+  targets = NULL,
+  assay = NULL,
+  group_vars = NULL,
+  mean_fxn = rowMeans,
+  fc_name = "difference",
+  p_adjust_method = c("bonferroni", "holm", "hochberg", "hommel", "BH", "BY", "fdr"),
+  verbose = TRUE,
+  ...
+) {
+  expect_Seurat()
+
+  # Validate input parameters
+  stopifnot(
+    "'contrast_column' must be available in Seurat object meta.data" =
+      contrast_column %in% colnames(object[[]])
+  )
+
+  .validate_da_input(object[[]], contrast_column, reference, targets, group_vars)
+
+  # Use default assay if assay = NULL
+  assay <- .validate_or_set_assay(object, assay)
+
+  # Define targets as all groups except the reference if not specified
+  targets <- targets %||% setdiff(unique(object[[]][, contrast_column, drop = TRUE]), reference)
+
+  # Check multiple choice args
+  p_adjust_method <- match.arg(p_adjust_method,
+    choices = c(
+      "bonferroni", "holm", "hochberg",
+      "hommel", "BH", "BY", "fdr"
+    )
+  )
+
+  # Add group data
+  group_data <- .select_group_data(object[[]], contrast_column, group_vars)
+
+  # Group data and get contrasts
+  if (verbose && check_global_verbosity() && !is.null(group_vars)) {
+    cli_alert_info("Splitting data by: {paste(c(group_vars), collapse = ', ')}")
+  }
+
+  test_groups <- group_data %>%
+    {
+      if (!is.null(group_vars)) {
+        group_by(., pick(all_of(group_vars)))
+      } else {
+        .
+      }
+    }
+
+  # Get group keys
+  test_groups_keys <- test_groups %>% group_keys()
+  if (nrow(test_groups_keys) == 1) {
+    group_data_split <- list(test_groups)
+  } else {
+    group_data_split <- test_groups %>%
+      group_split()
+  }
+
+  if (verbose && check_global_verbosity() && !is.null(group_vars)) {
+    .print_da_group_strategy(test_groups_keys, targets, reference)
+  }
+
+  # Get assay and convert if necessary
+  cg_assay <- object[[assay]]
+  if (inherits(cg_assay, "CellGraphAssay")) {
+    cg_assay <- as(cg_assay, "Assay")
+  }
+  if (inherits(cg_assay, "CellGraphAssay5")) {
+    cg_assay <- as(cg_assay, "Assay5")
+  }
+
+  de_results_all <- lapply(seq_along(group_data_split), function(i) {
```

Comment:
Really minor comment, but I would maybe rename "de" to "da" since it is protein abundance rather than gene expression.

Reply:
Makes sense, thanks!